flexmeasures/data/models/reporting/__init__.py

```diff
@@ -0,0 +1,184 @@
+    schema = ReporterConfigSchema
+    data: Dict[str, Union[tb.BeliefsDataFrame, pd.DataFrame]] = None
+    start: datetime = None
+    end: datetime = None
+    input_resolution: timedelta = None
+    belief_time: datetime = None
+
+    def __init__(
+        self, sensor: Sensor, reporter_config_raw: Optional[dict] = None
+    ) -> None:
+        """
+        Initialize a new Reporter.
+
+        Attributes:
+        :param sensor: sensor where the output of the reporter will be saved to.
+        :param reporter_config_raw: unserialized configuration of the reporter.
+        """
+
+        self.sensor = sensor
+
+        if not reporter_config_raw:
+            reporter_config_raw = {}
+
+        self.reporter_config_raw = reporter_config_raw
+
+    def fetch_data(self):
+        """
+        Fetches the time_beliefs from the database
+        """
+
+        self.data = {}
+        for tb_query in self.tb_query_config:
+            _tb_query = tb_query.copy()
+            # using start / end instead of event_starts_after/event_ends_before when not defined
+            event_starts_after = _tb_query.pop("event_starts_after", self.start)
+            event_ends_before = _tb_query.pop("event_ends_before", self.end)
+            resolution = _tb_query.pop("resolution", self.input_resolution)
+            belief_time = _tb_query.pop("belief_time", self.belief_time)
+
+            sensor: Sensor = _tb_query.pop("sensor", None)
+            alias: str = _tb_query.pop("alias", None)
+
+            bdf = sensor.search_beliefs(
+                event_starts_after=event_starts_after,
+                event_ends_before=event_ends_before,
+                resolution=resolution,
+                beliefs_before=belief_time,
+                **_tb_query,
+            )
+
+            # store data source as local variable
+            for source in bdf.sources.unique():
+                self.data[f"source_{source.id}"] = source
+
+            # store BeliefsDataFrame as local variable
+            if alias:
+                self.data[alias] = bdf
+            else:
+                self.data[f"sensor_{sensor.id}"] = bdf
+
+    def update_attribute(self, attribute, default):
+        if default is not None:
+            setattr(self, attribute, default)
+
+    def compute(
+        self,
+        *args,
+        start: datetime = None,
+        end: datetime = None,
+        input_resolution: timedelta = None,
+        belief_time: datetime = None,
+        **kwargs,
+    ) -> tb.BeliefsDataFrame:
+        """This method triggers the creation of a new report.
+
+        The same object can generate multiple reports with different start, end, input_resolution
+        and belief_time values.
+
+        In the future, this function will parse arbitrary input arguments defined in a schema.
+        """
+
+        # deserialize configuration
+        if self.reporter_config is None:
+            self.deserialize_config()
```

Comment:
Now that we deserialize this only once, does it still make sense that the `ReporterConfigSchema` contains query parameters?

Reply:
Not sure...

I like the idea of the reproducibility of having `start`, `end`, etc in a single json object. However, I'm open to remove them (in pro of simplifying the report_config). Let me know what you think.

Reply:
I think of the `reporter_config` (again, the -er matters) as something that configures a reporter once, so the reporter can be reused to generate different reports, with different time windows and/or other query parameters.

If you want the reporter to keep around (as a property) a dict object that includes start and end, I would call that a `report_config` (no -er), but it would be overwritten every time `compute` is called (perhaps `most_recent_report_config` might be a useful property name). But until we somehow use such a `report_config` for deriving a data source, I don't see how it serves a purpose other than stubbing that very idea in code.

If you want a reporter to be initializable with a `report_config` (no -er), I kind of understand that, but also think it's confusing.. what we'd really want it to initialize is a `Report` (which could be a class, but as we have BeliefsDataFrames for that, I don't see an immediate need to create a separate class). Perhaps the `Reporter` class should not even have `start` and `end` properties, because it is confusing the concept of reporters and reports.

The concept of a `report_config` maps better to the kwargs of the `add_report` (CLI) method you are working on.

Reply:
haha very true indeed, I was confusing `report_config` for `reporter_config`.

Good point. I'll remove the global time properties (`start`, `end`, ...) both from class attributes and the `ReporterConfigSchema`.